cpu/stm32l0/periph/i2c.c

```diff
@@ -0,0 +1,467 @@
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @addtogroup  driver_periph
+ * @{
+ *
+ * @file
+ * @brief       Low-level I2C driver implementation
+ *
+ * @note This implementation only implements the 7-bit addressing polling mode (for now
+ * interrupt mode is not available)
+ *
+ * @author      Aurélien Fillau <aurelien.fillau@we-sens.com>
+ *
+ * @}
+ */
+
+#include <stdint.h>
+
+
+#include "cpu.h"
+#include "mutex.h"
+#include "periph/i2c.h"
+#include "periph/gpio.h"
+#include "periph_conf.h"
+
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+
+/* guard file in case no I2C device is defined */
+#if I2C_NUMOF
+
+/* static function definitions */
+static void _i2c_init(I2C_TypeDef *i2c, uint32_t presc, uint32_t scll,
+                      uint32_t sclh, uint32_t sdadel, uint32_t scldel,
+                      uint32_t timing);
+static void _start(I2C_TypeDef *dev, uint8_t address, uint8_t length, uint8_t rw_flag);
+static inline void _read(I2C_TypeDef *dev, uint8_t *data, int length);
+static inline void _write(I2C_TypeDef *i2c, const uint8_t *data, int length);
+static inline void _stop(I2C_TypeDef *i2c);
+
+/**
+ * @brief Array holding one pre-initialized mutex for each I2C device
+ */
+static mutex_t locks[] =  {
+#if I2C_0_EN
+    [I2C_0] = MUTEX_INIT,
+#endif
+#if I2C_1_EN
+    [I2C_1] = MUTEX_INIT,
+#endif
+#if I2C_2_EN
+    [I2C_2] = MUTEX_INIT,
+#endif
+#if I2C_3_EN
+    [I2C_3] = MUTEX_INIT
+#endif
+};
+
+int i2c_init_master(i2c_t dev, i2c_speed_t speed)
+{
+    I2C_TypeDef *i2c;
+    gpio_t scl;
+    gpio_t sda;
+    uint32_t presc, scll, sclh, sdadel, scldel, timing;
+
+    /*
+     * Speed configuration:
+     * Example values can be found in the STM32L0xx Reference manual RM0367
+     * with f_I2CCLK = 16MHz
+     * t_SCLL: SCL low level counter
+     * t_SCLH: SCL high level counter
+     * t_SDADEL: Delay before sending SDA output
+     * t_SCLDEL: SCL low level during setup-time
+     */
+    switch (speed) {
+        case I2C_SPEED_NORMAL:
+            presc = 1;
+            scll = 0x56;  /* t_SCLL   = 5.0us */
+            sclh = 0x3E;   /* t_SCLH   = 4.0us */
+            sdadel = 0x1; /* t_SDADEL = 500ns */
+            scldel = 0xA; /* t_SCLDEL = 1250ns */
+            break;
+
+        case I2C_SPEED_FAST:
+            presc = 0;
+            scll = 0x2E;   /* t_SCLL   = 1250ns */
+            sclh = 0x11;   /* t_SCLH   = 500ns */
+            sdadel = 0x1; /* t_SDADEL = 125ns */
+            scldel = 0xB; /* t_SCLDEL = 500ns */
+            break;
+
+        case I2C_SPEED_FAST_PLUS:
+            presc = 0;
+            scll = 0x6;   /* t_SCLL   = 875ns */
+            sclh = 0x3;   /* t_SCLH   = 500ns */
+            sdadel = 0x0; /* t_SDADEL = 0ns */
+            scldel = 0x1; /* t_SCLDEL = 250ns */
+            break;
+
+        default:
+            return -2;
```

Comment:
There is no enum, the "-1" ad "-2" returned values are specified in the doxygen of the function.